File: lib/vajiram.js
/**
 * Vajiram & Ravi 100-Q Converter
 * Ported from the original HTML/JS to Node.js (Vercel serverless).
 * Uses two-column spatial parsing.
 */

const { getDocument } = require('pdfjs-dist');

async function extractPages(buffer) {
  const pdf = await getDocument({ data: buffer }).promise;
  const pages = [];
  for (let p = 1; p <= pdf.numPages; p++) {
    const page = await pdf.getPage(p);
    const vp = page.getViewport({ scale: 1.0 });
    const tc = await page.getTextContent();
    const items = tc.items
      .filter(i => i.str && i.str.trim())
      .map(i => ({
        text: i.str,
        x: Math.round(i.transform[4]),
        y: Math.round(vp.height - i.transform[5]),
        w: Math.round(i.width),
        h: Math.round(i.height || 10),
      }));
    pages.push({ items, width: vp.width, height: vp.height, pageNum: p });
  }
  return pages;
}

function consolidateSpatialTokens(tokensList, yTolerance = 4) {
  if (!tokensList.length) return [];
  const sorted = [...tokensList].sort((a, b) => a.y - b.y || a.x - b.x);
  const rows = [];
  let curRow = [sorted[0]];
  for (let i = 1; i < sorted.length; i++) {
    if (Math.abs(sorted[i].y - curRow[0].y) <= yTolerance) {
      curRow.push(sorted[i]);
    } else {
      rows.push(curRow);
      curRow = [sorted[i]];
    }
  }
  rows.push(curRow);
  return rows.map(bucket => {
    bucket.sort((a, b) => a.x - b.x);
    const text = bucket.map(t => t.text).join(' ').replace(/\s{2,}/g, ' ').trim();
    return { y: bucket[0].y, x: Math.min(...bucket.map(t => t.x)), text };
  }).filter(r => r.text.length > 0);
}

function isHeaderFooterNoise(line) {
  const s = line.trim().toUpperCase();
  if (s.length === 0) return true;
  const patterns = [
    /VAJIRAM\s*(&|AND)\s*RAVI/i, /PRELIMS\s*TEST\s*SERIES/i,
    /FULL\s*LENGTH\s*TEST/i, /TEST\s*BOOKLET/i,
    /MAXIMUM\s*MARKS/i, /TIME\s*ALLOWED/i, /DO\s*NOT\s*OPEN/i,
    /COMMENCEMENT\s*OF\s*THE\s*EXAMINATION/i, /UNPRINTED\s*OR\s*TORN/i,
    /CANDIDATE'S\s*RESPONSIBILITY/i, /ROLL\s*NUMBER/i,
    /OMR\s*ANSWER/i, /ANSWER\s*SHEET/i, /PENALTY\s*FOR\s*WRONG/i,
    /WRONG\s*ANSWERS\s*MARKED/i, /ALTERNATIVES\s*FOR\s*THE\s*ANSWER/i,
    /QUESTION\s*IS\s*LEFT\s*BLANK/i, /ECONOMICS\s*\(V\d+\)/i,
    /SCIENCE\s*&\s*TECHNOLOGY\s*\(V\d+\)/i, /POLITY\s*\(V\d+\)/i,
    /GS\s*TEST\s*-\s*\d+/i, /POWERUP\s*PRELIMS/i,
    /^\d{1,3}$/
  ];
  return patterns.some(p => p.test(line));
}

function isCoverPage(lines) {
  const combined = lines.map(l => l.text).join(' ');
  return !(/\(a\)/i.test(combined) || /\(b\)/i.test(combined));
}

function parseTestBooklet(pages, logFn) {
  const questionMap = {};
  for (const page of pages) {
    const masterLines = consolidateSpatialTokens(page.items);
    const filteredLines = masterLines.filter(l => !isHeaderFooterNoise(l.text));
    if (isCoverPage(filteredLines)) {
      if (logFn) logFn(`  Skipping non-question page: ${page.pageNum}`, 'warn');
      continue;
    }

    const midX = page.width / 2;
    const leftItems = page.items.filter(t => t.x < midX - 15);
    const rightItems = page.items.filter(t => t.x >= midX - 15);
    const isTwoColumn = leftItems.length > 6 && rightItems.length > 6;

    let unifiedText = '';
    if (isTwoColumn) {
      const leftLines = consolidateSpatialTokens(leftItems).filter(l => !isHeaderFooterNoise(l.text));
      const rightLines = consolidateSpatialTokens(rightItems).filter(l => !isHeaderFooterNoise(l.text));
      unifiedText = [...leftLines, ...rightLines].map(l => l.text).join('\n');
    } else {
      unifiedText = filteredLines.map(l => l.text).join('\n');
    }

    parseQuestionsFromText(unifiedText, questionMap);
  }
  return questionMap;
}

function parseQuestionsFromText(textStream, targetMap) {
  const lines = textStream.split('\n').map(r => r.trim()).filter(r => r.length > 0);
  let activeQ = null;
  let bodyLines = [];
  let options = [];
  let inOptions = false;

  function flush() {
    if (activeQ === null) return;
    if (options.length >= 2) {
      if (!targetMap[activeQ] || options.length > targetMap[activeQ].options.length) {
        targetMap[activeQ] = { id: activeQ, body: [...bodyLines], options: [...options] };
      }
    }
    activeQ = null;
    bodyLines = [];
    options = [];
    inOptions = false;
  }

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const optMatch = line.match(/^\s*\(([a-d])\)\s+(.+)$/i);
    if (optMatch && activeQ !== null) {
      inOptions = true;
      options.push({ letter: optMatch[1].toLowerCase(), text: optMatch[2].trim() });
      continue;
    }

    const qMatch = line.match(/^\s*(\d{1,3})\.\s{1,6}(.+)$/);
    if (qMatch) {
      const num = parseInt(qMatch[1], 10);
      if (num >= 1 && num <= 100) {
        const isListItem = (
          activeQ !== null && !inOptions &&
          num !== activeQ + 1 && num <= 6
        );
        if (!isListItem || activeQ === null) {
          flush();
          activeQ = num;
          bodyLines = [qMatch[2].trim()];
          options = [];
          inOptions = false;
          continue;
        }
      }
    }

    if (inOptions && options.length > 0 && activeQ !== null) {
      if (!line.match(/^\s*\(([a-d])\)/i)) {
        options[options.length - 1].text += ' ' + line;
      }
      continue;
    }

    if (activeQ !== null && !inOptions) {
      bodyLines.push(line);
    }
  }
  flush();
}

function parseSolutions(pages, logFn) {
  const allLines = [];
  for (const page of pages) {
    const formatted = consolidateSpatialTokens(page.items)
      .filter(l => !isHeaderFooterNoise(l.text))
      .map(l => l.text);
    allLines.push(...formatted);
  }

  const fullText = allLines.join('\n');

  // Extract answer keys
  const answerKeys = {};
  const keyRegex = /\b(\d{1,3})\.\s*\(([a-d])\)/gi;
  let m;
  while ((m = keyRegex.exec(fullText)) !== null) {
    const num = parseInt(m[1], 10);
    if (num >= 1 && num <= 100) {
      answerKeys[num] = m[2].toLowerCase();
    }
  }

  if (logFn) logFn(`  Answer keys found: ${Object.keys(answerKeys).length}`);

  // Extract explanations
  const explanations = {};
  const blockRegex = /\nQ(\d{1,3})\.\s*\n/g;
  const blocks = [];
  while ((m = blockRegex.exec(fullText)) !== null) {
    blocks.push({ num: parseInt(m[1], 10), start: m.index + m[0].length });
  }

  for (let i = 0; i < blocks.length; i++) {
    const { num, start } = blocks[i];
    const end = i + 1 < blocks.length ? blocks[i + 1].start : fullText.length;
    const slice = fullText.slice(start, end);
    explanations[num] = cleanExplanation(slice);
  }

  // Fallback: if no blocks found, try inline
  if (blocks.length < 5) {
    const inlineRegex = /\bQ\s*(\d{1,3})\s*\.\s*([A-D])\b/g;
    while ((m = inlineRegex.exec(fullText)) !== null) {
      const num = parseInt(m[1], 10);
      if (num >= 1 && num <= 100 && !answerKeys[num]) {
        answerKeys[num] = m[2].toLowerCase();
      }
    }
  }

  return { answerKeys, explanations };
}

function cleanExplanation(raw) {
  let t = raw;
  t = t.replace(/^Answer\s*:\s*[a-d]\s*$/gmi, '');
  t = t.replace(/^Explanation\s*:\s*$/gmi, '');
  t = t.replace(/Therefore[,\s]+option\s*\([a-d]\)\s*is\s*the\s*correct\s*answer\.?[^\n]*/gi, '');
  t = t.replace(/So[,\s]+option\s*\([a-d]\)\s*is\s*the\s*correct\s*answer\.?[^\n]*/gi, '');
  t = t.replace(/Therefore[,\s]+the\s*correct\s*answer[^\n]*/gi, '');
  t = t.replace(/Relevance\s*:[^\n]*/gi, '');
  t = t.replace(/^(?:Source|Ref|Reference)\s*:[^\n]*/gmi, '');
  t = t.replace(/^[\s]*[●○•▪◆▸▹→\-–—]+\s*/gm, '');
  t = t.replace(/^Q\d{1,3}\.\s*/gm, '');
  const cleaned = t.split('\n').map(l => l.trim()).filter(l => l.length > 2);
  return cleaned.join(' ').replace(/\s{2,}/g, ' ').replace(/\.\s*\./g, '.').trim();
}

function compileQuestionCoreLines(rawLines) {
  const normalized = rawLines.map(l => normalizeRoman(l.trim())).filter(l => l.length > 0);
  if (!normalized.length) return [];

  const newLineRe = [
    /^\d{1,2}\.\s+\S/,
    /^Statement\s+[IVXLC]+\s*:/i,
    /^(Which|How\s+many|How\s+|What|Select|Arrange|In\s+how|Who\s+|Where\s+|Among\s+|Identify|Of\s+the|With\s+reference|With\s+regard|Consider|Regarding|As\s+per|According\s+to|In\s+which\s+of\s+the\s+above)/i,
  ];

  const output = [];
  let buf = '';
  for (let i = 0; i < normalized.length; i++) {
    const line = normalized[i];
    const isNew = i === 0 || newLineRe.some(p => p.test(line));
    if (isNew) {
      if (buf) output.push(buf.replace(/\s{2,}/g, ' ').trim());
      buf = line;
    } else {
      buf += ' ' + line;
    }
  }
  if (buf) output.push(buf.replace(/\s{2,}/g, ' ').trim());
  return output;
}

function normalizeRoman(line) {
  return line.replace(
    /^\s*(I{1,3}|IV|V?I{0,3}|IX|XI{0,3})\.\s+/,
    (match, roman) => {
      const map = { I:1, II:2, III:3, IV:4, V:5, VI:6, VII:7, VIII:8, IX:9, X:10, XI:11, XII:12 };
      const digit = map[roman.toUpperCase()];
      return digit ? digit + '. ' : match;
    }
  );
}

function unpackOptions(rawOptions) {
  const combined = rawOptions.map(o => `(${o.letter}) ${o.text}`).join(' ');
  const matchA = combined.match(/\(a\)\s*([\s\S]*?)(?=\s*\(b\)|$)/i);
  const matchB = combined.match(/\(b\)\s*([\s\S]*?)(?=\s*\(c\)|$)/i);
  const matchC = combined.match(/\(c\)\s*([\s\S]*?)(?=\s*\(d\)|$)/i);
  const matchD = combined.match(/\(d\)\s*([\s\S]*?)$/i);
  const textA = matchA ? matchA[1] : 'Only one';
  const textB = matchB ? matchB[1] : 'Only two';
  const textC = matchC ? matchC[1] : 'Only three';
  const textD = matchD ? matchD[1] : 'All the four';
  const pure = s => s.replace(/^\s*\(?[a-d]\)?\s*\.?\s*/i, '').trim();
  return [
    { letter: 'a', text: pure(textA) },
    { letter: 'b', text: pure(textB) },
    { letter: 'c', text: pure(textC) },
    { letter: 'd', text: pure(textD) },
  ];
}

async function processVajiram(testBuffer, solBuffer, onProgress) {
  try {
    if (onProgress) onProgress(15, 'Decoding Test Booklet...');

    const testPages = await extractPages(testBuffer);
    if (onProgress) onProgress(40, 'Compiling questions...');

    const questionMap = parseTestBooklet(testPages);
    const qCount = Object.keys(questionMap).length;
    if (onProgress) onProgress(45, `Parsed ${qCount} questions from test booklet`);

    if (qCount === 0) {
      return {
        success: false,
        error: 'Zero valid questions found in Test PDF. Please check that it is a valid Vajiram test booklet with (a)(b)(c)(d) options.',
      };
    }

    if (onProgress) onProgress(60, 'Decoding Solution Booklet...');

    const solPages = await extractPages(solBuffer);
    if (onProgress) onProgress(85, 'Cross-referencing answers...');

    const { answerKeys, explanations } = parseSolutions(solPages, (msg) => {
      if (onProgress) onProgress(85 + (onProgress._inc || 0), msg);
    });

    if (onProgress) onProgress(95, 'Building output...');

    const sortedKeys = Object.keys(questionMap).map(Number).sort((a, b) => a - b);
    const lines = [];

    for (const key of sortedKeys) {
      const q = questionMap[key];
      const ansLetter = answerKeys[key];
      const expl = explanations[key] || 'Explanation not found for this question.';

      const coreLines = compileQuestionCoreLines(q.body);
      lines.push(`Q${key}. ${coreLines[0] || ''}`);
      for (let i = 1; i < coreLines.length; i++) lines.push(coreLines[i]);

      lines.push('😂');

      const opts = unpackOptions(q.options);
      for (const opt of opts) {
        const mark = (ansLetter && opt.letter === ansLetter) ? ' ✅' : '';
        lines.push(opt.text + mark);
      }

      lines.push(`Ex: ${expl.trim()}`);
      lines.push('');
    }

    const output = lines.join('\n');
    if (onProgress) onProgress(100, 'Done!');

    return {
      success: true,
      output,
      questionCount: qCount,
      lineCount: lines.length,
    };
  } catch (err) {
    return {
      success: false,
      error: err.message || String(err),
    };
  }
}

module.exports = { processVajiram };

File: bot/parsers/vajiram.py
"""
Vajiram & Ravi 100Q parser.
Requires two PDFs: Test Booklet + Solutions/Answer Key.
Handles 2-column spatial layout in test PDF.
"""
import re
import pdfplumber
from io import BytesIO

# Only match clear header/footer lines — DO NOT filter content words
NOISE_VAJ = [
    re.compile(r'VAJIRAM\s*(&|AND)\s*RAVI', re.I),
    re.compile(r'PRELIMS\s*TEST\s*SERIES', re.I),
    re.compile(r'FULL\s*LENGTH\s*TEST', re.I),
    re.compile(r'MAXIMUM\s*MARKS\s*:', re.I),
    re.compile(r'TIME\s*ALLOWED\s*:', re.I),
    re.compile(r'DO\s*NOT\s*OPEN\s*THIS', re.I),
    re.compile(r'COMMENCEMENT\s*OF\s*THE\s*EXAMINATION', re.I),
    re.compile(r"CANDIDATE.S\s*RESPONSIBILITY", re.I),
    re.compile(r'ROLL\s*NUMBER\s*:', re.I),
    re.compile(r'OMR\s*ANSWER\s*SHEET', re.I),
    re.compile(r'PENALTY\s*FOR\s*WRONG\s*ANSWER', re.I),
    re.compile(r'WRONG\s*ANSWERS?\s*MARKED', re.I),
    re.compile(r'GS\s*TEST\s*[-–]\s*\d+\s*[-–]', re.I),   # "GS Test - 11 -"
    re.compile(r'PowerUp\s*Prelims', re.I),
    re.compile(r'UNPRINTED\s*OR\s*TORN', re.I),
    re.compile(r'ALTERNATIVES?\s*FOR\s*THE\s*ANSWER', re.I),
    re.compile(r'QUESTION\s*IS\s*LEFT\s*BLANK', re.I),
    re.compile(r'^\s*\d{1,3}\s*$'),   # standalone page number only
]

ROMAN_MAP = {
    'I': 1, 'II': 2, 'III': 3, 'IV': 4, 'V': 5,
    'VI': 6, 'VII': 7, 'VIII': 8, 'IX': 9, 'X': 10,
    'XI': 11, 'XII': 12
}

NEW_LINE_RE = [
    re.compile(r'^\d{1,2}\.\s+\S'),
    re.compile(r'^Statement\s+[IVXLC]+\s*:', re.I),
    re.compile(
        r'^(Which|How\s+many|What|Select|Arrange|In\s+how|Who\b|'
        r'Where\b|Among\s+|Identify|Of\s+the|With\s+reference|'
        r'With\s+regard|Consider|Regarding|As\s+per|According\s+to|'
        r'In\s+which\s+of)', re.I
    ),
]


def is_noise_vaj(line: str) -> bool:
    s = line.strip()
    if not s:
        return True
    return any(p.search(s) for p in NOISE_VAJ)


def normalize_roman(line: str) -> str:
    def _rep(m):
        key = m.group(1).upper()
        n = ROMAN_MAP.get(key)
        return (str(n) + '. ') if n else m.group(0)
    return re.sub(r'^\s*(I{1,3}|IV|V?I{0,3}|IX|XI{0,3})\.\s+', _rep, line)


# ── PDF extraction ─────────────────────────────────────────────────────────────

def _extract_pages(pdf_bytes: bytes) -> list:
    pages = []
    with pdfplumber.open(BytesIO(pdf_bytes)) as pdf:
        for page in pdf.pages:
            words = page.extract_words(x_tolerance=3, y_tolerance=3)
            items = [{'text': w['text'], 'x': w['x0'], 'y': w['top']} for w in words]
            pages.append({
                'items': items,
                'width': float(page.width),
                'height': float(page.height),
                'page_id': page.page_number
            })
    return pages


def _items_to_lines(items: list, y_tol: int = 4) -> list:
    if not items:
        return []
    srt = sorted(items, key=lambda t: (t['y'], t['x']))
    buckets, cur = [], [srt[0]]
    for tok in srt[1:]:
        if abs(tok['y'] - cur[0]['y']) <= y_tol:
            cur.append(tok)
        else:
            buckets.append(cur)
            cur = [tok]
    buckets.append(cur)
    result = []
    for bkt in buckets:
        bkt.sort(key=lambda t: t['x'])
        txt = ' '.join(t['text'] for t in bkt).strip()
        if txt:
            result.append({'y': bkt[0]['y'], 'x': min(t['x'] for t in bkt), 'text': txt})
    return result


def _is_content_page(lines: list) -> bool:
    """Return True if page has actual question content (has (a)/(b) options)."""
    combined = ' '.join(l['text'] for l in lines)
    return bool(re.search(r'\(\s*[abcd]\s*\)', combined, re.I))


# ── Test booklet parser ────────────────────────────────────────────────────────

def _parse_test_booklet(pages: list) -> dict:
    q_map = {}
    for page in pages:
        all_lines = _items_to_lines(page['items'])
        filtered = [l for l in all_lines if not is_noise_vaj(l['text'])]

        if not _is_content_page(filtered):
            continue

        mid_x = page['width'] / 2
        left_items = [t for t in page['items'] if t['x'] < mid_x - 15]
        right_items = [t for t in page['items'] if t['x'] >= mid_x - 15]
        has_two_cols = len(left_items) > 6 and len(right_items) > 6

        if has_two_cols:
            left_lines = [l for l in _items_to_lines(left_items) if not is_noise_vaj(l['text'])]
            right_lines = [l for l in _items_to_lines(right_items) if not is_noise_vaj(l['text'])]
            text = ('\n'.join(l['text'] for l in left_lines) + '\n' +
                    '\n'.join(l['text'] for l in right_lines))
        else:
            text = '\n'.join(l['text'] for l in filtered)

        _parse_text_into_map(text, q_map)

    return q_map


def _parse_text_into_map(text: str, q_map: dict):
    """State-machine parser: extract question body + options from raw text."""
    lines = [l.strip() for l in text.split('\n') if l.strip()]
    active_id = None
    body = []
    opts = []
    in_opts = False

    def flush():
        nonlocal active_id, body, opts, in_opts
        if active_id is None:
            return
        if len(opts) >= 2:
            if active_id not in q_map or len(opts) > len(q_map[active_id]['options']):
                q_map[active_id] = {
                    'id': active_id,
                    'body_lines': list(body),
                    'options': list(opts)
                }
        active_id = None
        body = []
        opts = []
        in_opts = False

    OPT_RE = re.compile(r'^\s*\(([a-d])\)\s+(.+)$', re.I)
    # Question line: "N. text..." — use \s+ to handle any spacing
    Q_RE = re.compile(r'^(\d{1,3})\.\s+(.+)$')

    for line in lines:
        opt_m = OPT_RE.match(line)
        if opt_m and active_id is not None:
            in_opts = True
            opts.append({'letter': opt_m.group(1).lower(), 'text': opt_m.group(2).strip()})
            continue

        q_m = Q_RE.match(line)
        if q_m:
            num = int(q_m.group(1))
            if 1 <= num <= 100:
                # Decide: new question OR statement inside current question body?
                # Rule: treat as statement if we're in body (not opts), have body lines,
                # AND num is small enough to plausibly be a list item (<= 10),
                # AND it's NOT the expected next question (num != active_id + 1)
                # EXCEPTION: if we have no body yet, always start new question
                is_stmt = (
                    active_id is not None and
                    not in_opts and
                    len(body) > 0 and
                    num <= 10 and
                    num != active_id + 1
                )
                # Extra check: if num > active_id, very likely new question
                # (don't let large table row numbers bleed through)
                if is_stmt and num > active_id:
                    # Only trust it's internal if num is small (1-6 typical for statements)
                    is_stmt = num <= 6

                if not is_stmt:
                    flush()
                    active_id = num
                    body = [q_m.group(2).strip()]
                    opts = []
                    in_opts = False
                    continue
                # else: fall through and add to body

        if in_opts and opts and active_id is not None:
            if not OPT_RE.match(line):
                opts[-1]['text'] += ' ' + line
            continue

        if active_id is not None and not in_opts:
            body.append(line)

    flush()


# ── Solutions parser ───────────────────────────────────────────────────────────

def _parse_solutions(pages: list) -> tuple:
    answers = {}
    explanations = {}

    all_text = '\n'.join(
        '\n'.join(l['text'] for l in _items_to_lines(p['items'])
                  if not is_noise_vaj(l['text']))
        for p in pages
    )

    # Answer key: "N. (letter)" pattern
    for m in re.finditer(r'\b(\d{1,3})\.\s*\(([a-d])\)', all_text, re.I):
        num = int(m.group(1))
        if 1 <= num <= 100:
            answers[num] = m.group(2).lower()

    # Explanations: split on "QN." blocks (Vajiram solution format)
    splits = []
    for m in re.finditer(r'\nQ(\d{1,3})\.\s*\n', all_text):
        splits.append({'num': int(m.group(1)), 'start': m.end()})

    if not splits:
        # Fallback: try "QN." without blank line requirement
        for m in re.finditer(r'\bQ(\d{1,3})\.\s*\n', all_text):
            splits.append({'num': int(m.group(1)), 'start': m.end()})

    for i, sp in enumerate(splits):
        end = splits[i + 1]['start'] if i + 1 < len(splits) else len(all_text)
        raw = all_text[sp['start']:end]
        explanations[sp['num']] = _clean_expl(raw)

    return answers, explanations


def _clean_expl(raw: str) -> str:
    t = raw
    t = re.sub(r'Therefore[,\s]+option\s*\([a-d]\)\s*is\s*the\s*correct\s*answer\.?[^\n]*',
               '', t, flags=re.I)
    t = re.sub(r'So[,\s]+option\s*\([a-d]\)\s*is\s*the\s*correct\s*answer\.?[^\n]*',
               '', t, flags=re.I)
    t = re.sub(r'Therefore[,\s]+the\s*correct\s*answer[^\n]*', '', t, flags=re.I)
    t = re.sub(r'Relevance\s*:[^\n]*', '', t, flags=re.I)
    t = re.sub(r'(?:Source|Ref|Reference)\s*:[^\n]*', '', t, flags=re.I | re.M)
    t = re.sub(r'^[\s]*[●○•▪◆▸▹→\-–—]+\s*', '', t, flags=re.M)
    t = re.sub(r'^Q\d{1,3}\.\s*', '', t, flags=re.M)
    lines = [l.strip() for l in t.split('\n') if len(l.strip()) > 2]
    flat = re.sub(r'\s{2,}', ' ', ' '.join(lines))
    return re.sub(r'\.\s*\.', '.', flat).strip()


# ── Body line compilation ──────────────────────────────────────────────────────

def _compile_body(raw_lines: list) -> list:
    """Join continuation lines and normalize Roman numerals."""
    normalized = [normalize_roman(l.strip()) for l in raw_lines if l.strip()]
    if not normalized:
        return []
    out, current = [], ''
    for i, line in enumerate(normalized):
        starts_new = i == 0 or any(p.match(line) for p in NEW_LINE_RE)
        if starts_new:
            if current:
                out.append(re.sub(r'\s{2,}', ' ', current).strip())
            current = line
        else:
            current = re.sub(r'\s{2,}', ' ', current + ' ' + line)
    if current:
        out.append(re.sub(r'\s{2,}', ' ', current).strip())
    return out


def _unpack_options(options: list) -> list:
    """Extract clean option text from parsed (letter, text) pairs."""
    unified = ' '.join(f'({o["letter"]}) {o["text"]}' for o in options)
    result = []
    for i, letter in enumerate(['a', 'b', 'c', 'd']):
        next_letter = chr(ord(letter) + 1) if letter != 'd' else None
        if next_letter:
            pat = rf'\({re.escape(letter)}\)\s*([\s\S]*?)(?=\s*\({re.escape(next_letter)}\)|$)'
        else:
            pat = rf'\({re.escape(letter)}\)\s*([\s\S]*)$'
        m = re.search(pat, unified, re.I)
        txt = m.group(1).strip() if m else ''
        txt = re.sub(r'^\s*\(?[a-d]\)?\s*\.?\s*', '', txt, flags=re.I).strip()
        result.append({'letter': letter, 'text': txt or f'Option {letter.upper()}'})
    return result


# ── Public API ─────────────────────────────────────────────────────────────────

def parse_vajiram(test_pdf_bytes: bytes, sol_pdf_bytes: bytes) -> str:
    """Parse Vajiram & Ravi test + solution PDFs. Returns formatted text."""
    test_pages = _extract_pages(test_pdf_bytes)
    q_map = _parse_test_booklet(test_pages)

    sol_pages = _extract_pages(sol_pdf_bytes)
    answers, explanations = _parse_solutions(sol_pages)

    output = []
    for num in sorted(q_map.keys()):
        q = q_map[num]
        ans = answers.get(num, '')
        expl = explanations.get(num, '').strip()

        body = _compile_body(q['body_lines'])
        output.append(f'Q{num}. {body[0]}' if body else f'Q{num}.')
        for line in body[1:]:
            output.append(line)
        output.append('😂')

        for opt in _unpack_options(q['options']):
            mark = ' ✅' if ans and opt['letter'] == ans else ''
            output.append(opt['text'] + mark)

        output.append(f'Ex: {expl}' if expl else f'Ex: [Explanation not extracted for Q{num}]')
        output.append('')

    return '\n'.join(output)
